feat: Syntax for environment types now works with generics

alexvitkov opened the pull request:
# Description

<!-- Thanks for taking the time to improve Noir! -->
<!-- Please fill out all fields marked with an asterisk (*). -->

## Problem\*

The syntax for closure environment types that I added in https://github.com/noir-lang/noir/pull/2357 is sub-optimal as it does not play well with generics. 

Resolves: https://github.com/noir-lang/noir/issues/2334


## Summary\*
- As suggested in https://github.com/noir-lang/noir/pull/2357#discussion_r1298487380 this PR changes it changed from `fn[envarg1, envarg2](arg1, arg2, ...) -> ret` to `fn[env_type](arg1, arg2, ...) -> ret` so that a named generic can be used an environment. 

- Additional error checking is also added, since the new syntax allows passing any type as an environment, but the only valid ones are Unit/NamedGeneric/Tuple.

- Added tests for generic closure environments 



## Documentation

- [ ] This PR requires documentation updates when merged.

  <!-- If checked, check one of the following: -->

  - [ ] I will submit a noir-lang/docs PR.

  <!-- Submit a PR on https://github.com/noir-lang/docs. Thank you! -->

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  <!-- List / highlight what should be documented. -->
  <!-- Dev Rel will reach out for clarifications when needed. Thank you! -->

## Additional Context

<!-- Supplement further information if applicable. -->

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

jfecher commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -92,6 +92,10 @@ pub struct Resolver<'a> {
     /// parameter for the lambda function.
     lambda_stack: Vec<LambdaContext>,
+
+    /// UnresolvedTypes currently don't have Spans, so if we hit an error while resolving a type
+    /// the best we can do is show the expression where the type is used
+    current_expr_span: Option<Span>,
```
If we move this check to where the parser parses a closure type, we'll have access to the span there

alexvitkov replied:
In the parser we can't tell apart NamedGenerics from struct names/type aliases.

jfecher reviewed (changes requested):
Looks good but if we move where the error is checked for we can get it a bit cleaner

jfecher commented:
After testing out this PR, I think it is too confusing to report the error in a different source location. Can you add source locations to UnresolvedTypes so that the error location is correct?

alexvitkov commented:
I've added spans to UnresolvedType, the error now looks like this:
```
error: asdkj is not a valid closure environment type
  ┌─ /home/vitkov/code/repos/metacraft-labs/noir/proj/src/main.nr:4:16
  │
4 │ fn bad(foo: fn[asdkj]() -> Field) {
  │                ----- Closure environment must be a tuple or unit type
```

I thought about submitting a new PR, but it turned out not to be that massive of a change, so I dropped it in here so we can get it done faster

jfecher commented on crates/noirc_frontend/src/ast/mod.rs:
```diff
@@ -61,6 +61,16 @@ pub enum UnresolvedType {
+
+    // The span is None in the cases where the User omitted a type:
+    //  fn Foo() {}  --- return type is UnresolvedType::Unit without a span
+    //  let x = 100; --- type is UnresolvedType::Unspecified without a span
```
Good comment 👍

jfecher commented on crates/noirc_frontend/src/ast/mod.rs:
```diff
@@ -149,13 +165,27 @@ impl std::fmt::Display for UnresolvedTypeExpression {
         }
     }
+
+    pub fn with_span(&self, span: Span) -> UnresolvedType {
+        UnresolvedType { typ: self.clone(), span: Some(span) }
```
```suggestion
    pub fn with_span(self, span: Span) -> UnresolvedType {
        UnresolvedType { typ: self, span: Some(span) }
```
If the method requires an owned self, the signature should reflect that. Looking at where this method was used it was causing a lot of unnecessary clones before.

alexvitkov commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -342,32 +345,50 @@ impl<'a> Resolver<'a> {
+                // unwrap() here is valid, because the only places we don't have a span are omitted types
+                // e.g. a function without return type implicitly has a spanless UnresolvedType::Unit return type
+                // To get an invalid env type, the user must explicitly specify the type, which will have a span
+                let env_span = env.span.unwrap();
```
Makes sense, I'll keep it in mind: https://github.com/noir-lang/noir/pull/2383/commits/09721517ed624f4d783351255e1c5ae5d6f6aa9c

jfecher commented on crates/noirc_frontend/src/hir/resolution/resolver.rs:
```diff
@@ -342,32 +345,50 @@ impl<'a> Resolver<'a> {
+                // expect() here is valid, because the only places we don't have a span are omitted types
+                // e.g. a function without return type implicitly has a spanless UnresolvedType::Unit return type
+                // To get an invalid env type, the user must explicitly specify the type, which will have a span
+                let env_span = env.span.expect("Invalid closure environment type");
```
Why does the environment span being absent imply an invalid closure environment type?

alexvitkov replied:
Better error message: https://github.com/noir-lang/noir/pull/2383/commits/f2a90163498c9f8a0d47202e3b5be85c655da889

jfecher reviewed (approved):
👍
